### PR TITLE
Add new file based functions

### PR DESCRIPTION
**Pull request by logicminds**
The main goal was to create a mkdir_p function.  However, this also required some other functions to support the goal so I added those functions as well. 

  * path_join()
  * file_separator()
  * dir_split()
  * mkdir_p()

path_join is probably going to be a popular function as it also takes a windows path and turns it into unix form in addition to just joining parts of a path together by the separator.  The path join function should make for better multi platform path manipulation.

### CODE REVIEW

**Comment by logicminds** [comment]
You can test out the functions [here](http://demo.puppet-debugger.com/play?content=extlib::path_join([%27c:\windows\temp%27])).

Also since these are native puppet functions, they only work with puppet 4.2+ I believe.

File: functions/dir_split.pp [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+        $acc_value = ($acc[$counter] =~ Undef) ? { true => '', false => $acc[$counter] }
+        unless empty($value) {
+          $acc + extlib::path_join([$acc_value, $value])
+        } else {
+          $acc
+        }
+      }
+  }
+  $dirs_array.flatten.unique
+}
```
Comment (alexjfisher) [inline functions/dir_split.pp]: nit.  Can you add new lines?

File: functions/dir_split.pp [hunk 2]
```diff
@@ -0,0 +1,19 @@
+# @param $dirs - either an absolute path or a array of absolute paths
+# @return Array[String] - an array of absolute paths after being cut into individual paths
```
Comment (alexjfisher) [inline functions/dir_split.pp]: The type information here shouldn't be needed?

**Comment by alexjfisher** [comment]
Looks good.  Could you regenerate the REFERENCE.MD (and the instructions for how to do this in the README)?

**Comment by logicminds** [comment]
> Looks good. Could you regenerate the REFERENCE.MD (and the instructions for how to do this in the README)?

When it said auto generated in Reference.  I assumed it was auto generated :)

File: Gemfile [hunk 3]
```diff
@@ -66,7 +66,7 @@ group :release do
   gem 'github_changelog_generator',  :require => false, :git => 'https://github.com/github-changelog-generator/github-changelog-generator' if RUBY_VERSION >= '2.2.2'
   gem 'puppet-blacksmith',           :require => false
   gem 'voxpupuli-release',           :require => false, :git => 'https://github.com/voxpupuli/voxpupuli-release-gem'
-  gem 'puppet-strings', '>= 1.0',    :require => false
+  gem 'puppet-strings', '>= 2.0',    :require => false
```
Comment (alexjfisher) [inline Gemfile]: Could you open a PR for this change [here](https://github.com/voxpupuli/modulesync_config/blob/master/config_defaults.yml#L93) instead?

**Comment by alexjfisher** [comment]
@logicminds I'm rebasing this now and will be merging shortly.

**Pull request merged**